File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression_me.R
# Polynomial Regression 

# data  preprocessing phase
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Fitting Linear Regression Model into dataset
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

# Fitting Polynomial Linear Regression Model into dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

lin_reg2 = lm(formula = Salary ~ ., 
               data = dataset)

library(ggplot2)
# Visualizing Linear Regression Model
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = "blue") + 
  ggtitle('Truth vs Bluff (Linear Regression)') + 
  xlab('Level') + 
  ylab('Salary')

# Visualizing Polynomial Linear Regression Model
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg2, newdata = dataset)),
            colour = "blue") + 
  ggtitle('Truth vs Bluff (Linear Regression)') + 
  xlab('Level') + 
  ylab('Salary')

# Visualizing Polynomial Linear Regression Model (High Resolution)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = "red") + 
  geom_line(aes(x = x_grid, y = predict(lin_reg2, newdata = data.frame(Level = x_grid,
                                                                       Level2 = x_grid^2,
                                                                       Level3 = x_grid^3,
                                                                       Level4 = x_grid^4))),
            colour = "blue") + 
  ggtitle('Truth vs Bluff (Linear Regression)') + 
  xlab('Level') + 
  ylab('Salary')


# predicting a new result with Linear Regression
predict(lin_reg, data.frame(Level = 6.5))

# prediction a new result with Polynomial Regression
predict(lin_reg2, data.frame(Level = 6.5,
                             Level2 = 6.5^2,
                             Level3 = 6.5^3,
                             Level4 = 6.5^4))

File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression_me.py
# Polynomial Regression

# data pre-processing phase
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# importing dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Fitting Simple Linear Regression Model to dataset
from sklearn.linear_model import LinearRegression
lin_regressor = LinearRegression()
lin_regressor.fit(X, y)

# Fitting Polynomial Linear Regression Model to dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)
poly_reg.fit(X_poly, y)
lin_regressor2 = LinearRegression()
lin_regressor2.fit(X_poly, y)

# Visualizing Linear Regression Results
plt.scatter(X, y, color = "red")
plt.plot(X, lin_regressor.predict(X), color = "blue")
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# Visualizing Polynomial Linear Regression Results
plt.scatter(X, y, color = "red")
plt.plot(X, lin_regressor2.predict(poly_reg.fit_transform(X)), color = "blue")
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

# Visualizing Polynomial Linear Regression Results (High Resolution)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = "red")
plt.plot(X_grid, lin_regressor2.predict(poly_reg.fit_transform(X_grid)), color = "blue")
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()
